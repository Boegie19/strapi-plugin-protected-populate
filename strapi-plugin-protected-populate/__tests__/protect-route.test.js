const { protectRoute } = require('../server/middlewares/helpers/protect-route');

describe('Protect populate', () => {
  describe('Normal fields', () => {
    it('should only populate what is allowed', () => {
      let query = { fields: ['a', 'b'] };
      info = { fields: ['b'] };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
    it('should return only the fields in info when fields does not exist', () => {
      let query = {};
      info = { fields: ['b'] };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
    it('should return full info when unexpected fields input', () => {
      let query = { fields: 't' };
      info = { fields: ['b'] };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
    it('should return no fields noting', () => {
      let query = { fields: ['a'] };
      info = {};
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
    it('should return correctly if nested in populate', () => {
      let query = { populate: { abc: { fields: ['a', 'b'] } } };
      info = { populate: { abc: { fields: ['b'] } } };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
  });
  describe('Normal populate', () => {
    it('populate should only populate if in info', () => {
      let query = { populate: { a: {}, b: {} } };
      info = { populate: { b: {} } };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
    it('populate should not populate anyting if info has no populate', () => {
      let query = { populate: { a: {}, b: {} } };
      info = {};
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
    it('populate should not populate anyting if quary has no populate', () => {
      let query = {};
      info = { populate: { a: {}, b: {} } };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
  });
  describe('Normal array populate', () => {
    it('populate should only populate should handle 3 populates', () => {
      let query = { populate: ['a.b.c'] };
      info = {
        populate: {
          a: {
            populate: {
              b: {
                populate: {
                  c: {
                    fields: ['a', 'b'],
                  },
                },
              },
            },
          },
        },
      };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
    it('populate should still protect in array', () => {
      let query = { populate: ['a.b.c'] };
      info = {
        populate: {
          a: {
            populate: {
              b: {},
            },
          },
        },
      };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
  });
  describe('Normal * populate', () => {
    it('populate * should only show first level down', () => {
      let query = { populate: '*' };
      info = {
        populate: {
          a: {
            fields: ['a', 'z'],
            populate: {
              b: {
                fields: ['a', 'b'],
              },
            },
          },
        },
      };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
  });
  describe('Normal object = true populate', () => {
    it('populate a: true should populate 1 level of a only', () => {
      let query = { populate: { a: true } };
      info = {
        populate: {
          a: {
            fields: ['a', 'z'],
            populate: {
              b: {
                fields: ['a', 'b'],
              },
            },
          },
          b: {
            fields: ['a', 'z'],
            populate: {
              b: {
                fields: ['a', 'b'],
              },
            },
          },
        },
      };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
  });

  describe('DZ populate fields', () => {
    it('populate DZ fields', () => {
      let query = { fields: ['c', 'f', 'z'] };
      info = {
        on: {
          'a.a': {
            fields: ['c', 'd'],
            populate: {
              c: {
                fields: ['a', 'b'],
              },
              d: {
                fields: ['a', 'b'],
              },
            },
          },
          'a.b': {
            fields: ['c', 'd'],
            populate: {
              c: {
                fields: ['a', 'b'],
              },
            },
          },
          'a.c': {
            fields: ['e', 'f'],
            populate: {
              c: {
                fields: ['a', 'b'],
              },
            },
          },
        },
      };

      expect(protectRoute(query, info)).toMatchSnapshot();
    });
  });
  describe('DZ populate', () => {
    it('should populate DZ and split populate between all components', () => {
      let query = {
        populate: {
          c: {},
          d: {},
        },
      };
      info = {
        on: {
          'a.a': {
            fields: ['c', 'd'],
            populate: {
              c: {
                fields: ['a', 'b'],
              },
              d: {
                fields: ['a', 'b'],
              },
            },
          },
          'a.b': {
            fields: ['c', 'd'],
            populate: {
              c: {
                fields: ['a', 'b'],
              },
            },
          },
          'a.c': {
            fields: ['e', 'f'],
            populate: {
              f: {
                fields: ['a', 'b'],
              },
            },
          },
        },
      };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
  });
  describe('DZ array populate', () => {
    it('should populate DZ with populate array', () => {
      let query = {
        populate: ['c', 'd'],
      };
      info = {
        on: {
          'a.a': {
            fields: ['c', 'd'],
            populate: {
              c: {
                fields: ['a', 'b'],
              },
              d: {
                fields: ['a', 'b'],
              },
            },
          },
          'a.b': {
            fields: ['c', 'd'],
            populate: {
              c: {
                fields: ['a', 'b'],
              },
            },
          },
          'a.c': {
            fields: ['e', 'f'],
            populate: {
              f: {
                fields: ['a', 'b'],
              },
            },
          },
        },
      };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
  });
  describe('DZ populate *', () => {
    it('should populate DZ *', () => {
      let query = {
        populate: '*',
      };
      info = {
        populate: {
          dz: {
            on: {
              'a.a': {
                fields: ['c', 'd'],
                populate: {
                  c: {
                    fields: ['a', 'b'],
                  },
                  d: {
                    fields: ['a', 'b'],
                  },
                },
              },
              'a.b': {
                fields: ['c', 'd'],
                populate: {
                  c: {
                    fields: ['a', 'b'],
                  },
                },
              },
              'a.c': {
                fields: ['e', 'f'],
                populate: {
                  f: {
                    fields: ['a', 'b'],
                  },
                },
              },
            },
          },
          abc: {
            fields: ['a', 'b'],
            populate: {
              f: {
                fields: ['a', 'b'],
              },
            },
          },
        },
      };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
  });
  describe('DZ object = true', () => {
    it('should populate DZ if DZ value is true', () => {
      let query = {
        populate: {
          dz: true,
        },
      };
      info = {
        populate: {
          dz: {
            on: {
              'a.a': {
                fields: ['c', 'd'],
                populate: {
                  c: {
                    fields: ['a', 'b'],
                  },
                  d: {
                    fields: ['a', 'b'],
                  },
                },
              },
              'a.b': {
                fields: ['c', 'd'],
                populate: {
                  c: {
                    fields: ['a', 'b'],
                  },
                },
              },
              'a.c': {
                fields: ['e', 'f'],
                populate: {
                  f: {
                    fields: ['a', 'b'],
                  },
                },
              },
            },
          },
          abc: {
            fields: ['a', 'b'],
            populate: {
              f: {
                fields: ['a', 'b'],
              },
            },
          },
        },
      };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
  });

  describe('DZ on', () => {
    it('should populate DZ name only if on is empty', () => {
      let query = {
        populate: {
          dz: {
            on: {},
          },
        },
      };
      info = {
        populate: {
          dz: {
            on: {
              'a.a': {
                fields: ['c', 'd'],
                populate: {
                  c: {
                    fields: ['a', 'b'],
                  },
                  d: {
                    fields: ['a', 'b'],
                  },
                },
              },
              'a.b': {
                fields: ['c', 'd'],
                populate: {
                  c: {
                    fields: ['a', 'b'],
                  },
                },
              },
              'a.c': {
                fields: ['e', 'f'],
                populate: {
                  f: {
                    fields: ['a', 'b'],
                  },
                },
              },
            },
          },
          abc: {
            fields: ['a', 'b'],
            populate: {
              f: {
                fields: ['a', 'b'],
              },
            },
          },
        },
      };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });

    it('should populate DZ only selected fields/populates', () => {
      let query = {
        populate: {
          dz: {
            on: {
              'a.a': {
                fields: ['c'],
                populate: {
                  c: {
                    fields: ['a'],
                  },
                },
              },
            },
          },
        },
      };
      info = {
        populate: {
          dz: {
            on: {
              'a.a': {
                fields: ['c', 'd'],
                populate: {
                  c: {
                    fields: ['a', 'b'],
                  },
                  d: {
                    fields: ['a', 'b'],
                  },
                },
              },
              'a.b': {
                fields: ['c', 'd'],
                populate: {
                  c: {
                    fields: ['a', 'b'],
                  },
                },
              },
              'a.c': {
                fields: ['e', 'f'],
                populate: {
                  f: {
                    fields: ['a', 'b'],
                  },
                },
              },
            },
          },
          abc: {
            fields: ['a', 'b'],
            populate: {
              f: {
                fields: ['a', 'b'],
              },
            },
          },
        },
      };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
  });
});

describe('Protect filters', () => {
  describe('Normal filters', () => {
    it('check if i can filter on a allowed field', () => {
      let query = { filters: { b: { $eq: 'a' } } };
      info = { fields: ['b'] };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
    it("ensure that I can't filter on none allowed fields", () => {
      let query = { filters: { a: { $eq: 'a' } } };
      info = { fields: ['b'] };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
  });
  describe('joins Filters', () => {
    it('should ignore $and', () => {
      let query = {
        filters: { $and: [{ a: { $eq: 'a' } }, { b: { $eq: 'b' } }, { c: { $eq: 'c' } }] },
      };
      info = { fields: ['a', 'b'] };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
    it('should ignore $or', () => {
      let query = {
        filters: { $or: [{ a: { $eq: 'a' } }, { b: { $eq: 'b' } }, { c: { $eq: 'c' } }] },
      };
      info = { fields: ['a', 'b'] };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
    it('should ignore $not and work', () => {
      let query = {
        filters: { $not: { a: { $eq: 'a' } } },
      };
      info = { fields: ['a', 'b'] };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
    it('should ignore $not and and ignore none existing fields', () => {
      let query = {
        filters: { $not: { c: { $eq: 'a' } } },
      };
      info = { fields: ['a', 'b'] };
      expect(protectRoute(query, info)).toMatchSnapshot();
    });
  });
});
