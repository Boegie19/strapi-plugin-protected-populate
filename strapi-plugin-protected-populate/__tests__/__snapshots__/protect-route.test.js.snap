// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Protect filters Normal filters check if i can filter on a allowed field 1`] = `
{
  "fields": [
    "id",
    "b",
  ],
  "filters": {
    "b": {
      "$eq": "a",
    },
  },
  "populate": {},
}
`;

exports[`Protect filters Normal filters ensure that I can't filter on none allowed fields 1`] = `
{
  "fields": [
    "id",
    "b",
  ],
  "filters": {},
  "populate": {},
}
`;

exports[`Protect filters joins Filters should ignore $and 1`] = `
{
  "fields": [
    "id",
    "a",
    "b",
  ],
  "filters": {
    "$and": [
      {
        "a": {
          "$eq": "a",
        },
      },
      {
        "b": {
          "$eq": "b",
        },
      },
      {},
    ],
  },
  "populate": {},
}
`;

exports[`Protect filters joins Filters should ignore $not and and ignore none existing fields 1`] = `
{
  "fields": [
    "id",
    "a",
    "b",
  ],
  "filters": {},
  "populate": {},
}
`;

exports[`Protect filters joins Filters should ignore $not and work 1`] = `
{
  "fields": [
    "id",
    "a",
    "b",
  ],
  "filters": {},
  "populate": {},
}
`;

exports[`Protect filters joins Filters should ignore $or 1`] = `
{
  "fields": [
    "id",
    "a",
    "b",
  ],
  "filters": {
    "$or": [
      {
        "a": {
          "$eq": "a",
        },
      },
      {
        "b": {
          "$eq": "b",
        },
      },
      {},
    ],
  },
  "populate": {},
}
`;

exports[`Protect populate DZ array populate should populate DZ with populate array 1`] = `
{
  "fields": undefined,
  "filters": undefined,
  "populate": {
    "a.a": {
      "fields": [
        "id",
        "c",
        "d",
      ],
      "populate": {
        "c": {
          "fields": [
            "id",
            "a",
            "b",
          ],
          "populate": {},
        },
        "d": {
          "fields": [
            "id",
            "a",
            "b",
          ],
          "populate": {},
        },
      },
    },
    "a.b": {
      "fields": [
        "id",
        "c",
        "d",
      ],
      "populate": {
        "c": {
          "fields": [
            "id",
            "a",
            "b",
          ],
          "populate": {},
        },
      },
    },
    "a.c": {
      "fields": [
        "id",
        "e",
        "f",
      ],
      "populate": {},
    },
  },
}
`;

exports[`Protect populate DZ object = true should populate DZ if DZ value is true 1`] = `
{
  "fields": [
    "id",
  ],
  "filters": undefined,
  "populate": {
    "dz": {
      "on": {
        "a.a": {
          "fields": [
            "id",
            "c",
            "d",
          ],
        },
        "a.b": {
          "fields": [
            "id",
            "c",
            "d",
          ],
        },
        "a.c": {
          "fields": [
            "id",
            "e",
            "f",
          ],
        },
      },
    },
  },
}
`;

exports[`Protect populate DZ on should populate DZ name only if on is empty 1`] = `
{
  "fields": [
    "id",
  ],
  "filters": undefined,
  "populate": {
    "dz": {
      "on": {},
    },
  },
}
`;

exports[`Protect populate DZ on should populate DZ only selected fields/populates 1`] = `
{
  "fields": [
    "id",
  ],
  "filters": undefined,
  "populate": {
    "dz": {
      "on": {
        "a.a": {
          "fields": [
            "id",
            "c",
          ],
          "populate": {
            "c": {
              "fields": [
                "id",
                "a",
              ],
              "populate": {},
            },
          },
        },
      },
    },
  },
}
`;

exports[`Protect populate DZ populate * should populate DZ * 1`] = `
{
  "fields": [
    "id",
  ],
  "filters": undefined,
  "populate": {
    "abc": {
      "fields": [
        "id",
        "a",
        "b",
      ],
    },
    "dz": {
      "on": {
        "a.a": {
          "fields": [
            "id",
            "c",
            "d",
          ],
        },
        "a.b": {
          "fields": [
            "id",
            "c",
            "d",
          ],
        },
        "a.c": {
          "fields": [
            "id",
            "e",
            "f",
          ],
        },
      },
    },
  },
}
`;

exports[`Protect populate DZ populate fields populate DZ fields 1`] = `
{
  "fields": undefined,
  "filters": undefined,
  "populate": {
    "a.a": {
      "fields": [
        "id",
        "c",
      ],
    },
    "a.b": {
      "fields": [
        "id",
        "c",
      ],
    },
    "a.c": {
      "fields": [
        "id",
        "f",
      ],
    },
  },
}
`;

exports[`Protect populate DZ populate should populate DZ and split populate between all components 1`] = `
{
  "fields": undefined,
  "filters": undefined,
  "populate": {
    "a.a": {
      "fields": [
        "id",
        "c",
        "d",
      ],
      "populate": {
        "c": {
          "fields": [
            "id",
            "a",
            "b",
          ],
          "populate": {},
        },
        "d": {
          "fields": [
            "id",
            "a",
            "b",
          ],
          "populate": {},
        },
      },
    },
    "a.b": {
      "fields": [
        "id",
        "c",
        "d",
      ],
      "populate": {
        "c": {
          "fields": [
            "id",
            "a",
            "b",
          ],
          "populate": {},
        },
      },
    },
    "a.c": {
      "fields": [
        "id",
        "e",
        "f",
      ],
      "populate": {},
    },
  },
}
`;

exports[`Protect populate Normal * populate populate * should only show first level down 1`] = `
{
  "fields": [
    "id",
  ],
  "filters": undefined,
  "populate": {
    "a": {
      "fields": [
        "id",
        "a",
        "z",
      ],
    },
  },
}
`;

exports[`Protect populate Normal array populate populate should only populate should handle 3 populates 1`] = `
{
  "fields": [
    "id",
  ],
  "filters": undefined,
  "populate": {
    "a": {
      "fields": [
        "id",
      ],
      "populate": {
        "b": {
          "fields": [
            "id",
          ],
          "populate": {
            "c": {
              "fields": [
                "id",
                "a",
                "b",
              ],
              "populate": {},
            },
          },
        },
      },
    },
  },
}
`;

exports[`Protect populate Normal array populate populate should still protect in array 1`] = `
{
  "fields": [
    "id",
  ],
  "filters": undefined,
  "populate": {
    "a": {
      "fields": [
        "id",
      ],
      "populate": {
        "b": {
          "fields": [
            "id",
          ],
          "populate": {},
        },
      },
    },
  },
}
`;

exports[`Protect populate Normal fields should only populate what is allowed 1`] = `
{
  "fields": [
    "id",
    "b",
  ],
  "filters": undefined,
  "populate": {},
}
`;

exports[`Protect populate Normal fields should return correctly if nested in populate 1`] = `
{
  "fields": [
    "id",
  ],
  "filters": undefined,
  "populate": {
    "abc": {
      "fields": [
        "id",
        "b",
      ],
      "populate": {},
    },
  },
}
`;

exports[`Protect populate Normal fields should return full info when unexpected fields input 1`] = `
{
  "fields": [
    "id",
    "b",
  ],
  "filters": undefined,
  "populate": {},
}
`;

exports[`Protect populate Normal fields should return no fields noting 1`] = `
{
  "fields": [
    "id",
  ],
  "filters": undefined,
  "populate": {},
}
`;

exports[`Protect populate Normal fields should return only the fields in info when fields does not exist 1`] = `
{
  "fields": [
    "id",
    "b",
  ],
  "filters": undefined,
  "populate": {},
}
`;

exports[`Protect populate Normal object = true populate populate a: true should populate 1 level of a only 1`] = `
{
  "fields": [
    "id",
  ],
  "filters": undefined,
  "populate": {
    "a": true,
  },
}
`;

exports[`Protect populate Normal populate populate should not populate anyting if info has no populate 1`] = `
{
  "fields": [
    "id",
  ],
  "filters": undefined,
  "populate": {},
}
`;

exports[`Protect populate Normal populate populate should not populate anyting if quary has no populate 1`] = `
{
  "fields": [
    "id",
  ],
  "filters": undefined,
  "populate": undefined,
}
`;

exports[`Protect populate Normal populate populate should only populate if in info 1`] = `
{
  "fields": [
    "id",
  ],
  "filters": undefined,
  "populate": {
    "b": {
      "fields": [
        "id",
      ],
      "populate": {},
    },
  },
}
`;
